quickwit/quickwit-actors/src/registry.rs

```diff
@@ -181,6 +223,39 @@ fn get_iter<A: Actor>(
         .filter(|mailbox| !mailbox.is_disconnected())
 }
 
+/// This structure contains an optional exit handle. The handle is present
+/// until the join() method is called.
+#[derive(Clone)]
+pub(crate) struct ActorJoinHandle {
+    holder: Arc<Mutex<Option<JoinHandle<ActorExitStatus>>>>,
+}
+
+impl ActorJoinHandle {
+    pub(crate) fn new(join_handle: JoinHandle<ActorExitStatus>) -> Self {
+        ActorJoinHandle {
+            holder: Arc::new(Mutex::new(Some(join_handle))),
+        }
+    }
+
+    /// Joins the actor and returns its exit status on the frist invocation.
+    /// Returns None afterwards.
+    pub(crate) async fn join(&self) -> Option<ActorExitStatus> {
+        let mut guard = self.holder.lock().await;
+        if let Some(join_handle) = guard.take() {
+            let exit_status = join_handle.await.unwrap_or_else(|join_err| {
+                if join_err.is_panic() {
+                    ActorExitStatus::Panicked
+                } else {
+                    ActorExitStatus::Killed
+                }
+            });
+            Some(exit_status)
+        } else {
+            None
+        }
+    }
```

Comment:
nitpick...

Whenever you hold a lock, you want to make it trivial for the reader that what you are doing is simple.

Make the code running while holding the lock as short and as simple as possible. Ideally not run any io or call external code.

Here I suggest we do it as follows

```suggestion
        // Lock is held only on the following line.
        let mut join_handle_opt = self.holder.lock().await.take();
        if let Some(join_handle) = join_handle {
            let exit_status = join_handle.await.unwrap_or_else(|join_err| {
                if join_err.is_panic() {
                    ActorExitStatus::Panicked
                } else {
                    ActorExitStatus::Killed
                }
            });
            Some(exit_status)
        } else {
            None
        }
    }
```

Reply:
Or maybe this is a bad idea actually...

With the following solution, a second person calling `.join()` concurrently will return `None` instantly which might not be what you wanted.

Reply:
Maybe we can do something cleaner using https://docs.rs/futures/latest/futures/future/trait.FutureExt.html#method.shared ?

Reply:
That's a decision leftover from one of the iterations where I was storing enum with an exit code here instead of Option and returning it to the second caller instead of panicking. In that solution it was important to wait for join to finish. I am not sure if it is important in the current setup. It "feels" like it is safer to block the second caller until join actually happens. But I don't really have any solid arguments at the moment for either approach.

Reply:
I suspected something like that. Can you have a look at the Shared<JoinHandle> solution?
It will simplify the code a little bit, and return the exit code in subsequent calls.